### PR TITLE
chore(amplify_api): support decoding custom list request

### PR DESCRIPTION
**Pull request by ragingsquirrel3**
## Issues

* https://github.com/aws-amplify/amplify-flutter/issues/816
* https://github.com/aws-amplify/amplify-flutter/issues/1425

Model helpers support the use case of manually creating a request with some metadata so that it will get decoded into the correct type (i.e. instance of a model rather than string/JSON). This was tested/documented for requests of a single instance (get, update, create, delete) but not for custom list queries. They require a certain class `PaginatedModelTypeImpl` (implementation of abstract class `PaginatedModelType` that is not publicly exported to use properly).

To fix that, this PR refactors class definitions and decoding logic related to paginated results. It gets rid of the abstract/implementation classes and instead has simple class definitions in `PaginatedModelType` and `PaginatedResult` which are already public. The result is that caller can use `PaginatedModelType` as the `modelType` for request and they don't need the old private version. It has some tests, too.

Generally, this is a simplification, however, it required moving a small bit of logic for getting the request for next page of results to the decoder.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: packages/amplify_api/example/integration_test/graphql_tests.dart [hunk 1]
```diff
@@ -313,6 +313,29 @@ void main() {
...
+            items {
+              id
+              name
+              createdAt
+            }
+          }
+        }''';
+        final request = GraphQLRequest<PaginatedResult<Blog>>(
+            document: graphQLDocument,
+            modelType: const PaginatedModelTypeImpl(Blog.classType),
```
Comment (ragingsquirrel3) [inline packages/amplify_api/example/integration_test/graphql_tests.dart:334]: Callout: when we named this class we did not intend for it to be developer-facing as we were not considering the customer request use case very much. Now that we are exposing it like this, don't know if we like a different name?

Comment (Jordan-Nelson) [reply packages/amplify_api/example/integration_test/graphql_tests.dart:334]: I don't have enough context to suggest a name, but I would vote for renaming it.

Are we expecting there to be other classes that extend `PaginatedModelType`? Do we need `PaginatedModelType` and `PaginatedModelTypeImpl`?

Comment (dnys1) [reply packages/amplify_api/example/integration_test/graphql_tests.dart:334]: What about:

```dart
extension PaginatedModelTypeExt<T extends Model> on ModelType<T> {
  PaginatedModelType<T> get list => PaginatedModelTypeImpl(this);
}

// so it becomes

GraphQLRequest<PaginatedResult<Blog>>(
  document: graphQLDocument,
  modelType: Blog.classType.list,
  // ...
```

This way the class can stay private and it's more in line with the current approach.

**Comment by ragingsquirrel3** [comment]
I think this fix will help with use case mentioned in https://github.com/aws-amplify/amplify-flutter/issues/816

File: packages/amplify_api/example/integration_test/graphql_tests.dart (hunk 1 above)
Comment (ragingsquirrel3) [reply packages/amplify_api/example/integration_test/graphql_tests.dart]: In latest revision refactored to follow @Jordan-Nelson 's suggestion here.

File: packages/amplify_api_plugin_interface/lib/src/types/pagination/paginated_result.dart [hunk 2]
```diff
@@ -17,21 +17,45 @@ import 'package:amplify_core/amplify_core.dart';
 
 import '../../types.dart';
 
-abstract class PaginatedResult<T extends Model> extends Model {
+class PaginatedResult<T extends Model> extends Model {
+  /// Model instances for this set of results. An entry might be null if
```
Comment (Jordan-Nelson) [inline packages/amplify_api_plugin_interface/lib/src/types/pagination/paginated_result.dart:21]: *nit:* https://dart.dev/guides/language/effective-dart/documentation#do-separate-the-first-sentence-of-a-doc-comment-into-its-own-paragraph

```suggestion
  /// Model instances for this set of results.
  ///
  /// An entry might be null if
```

**Review by Jordan-Nelson — APPROVED** [review]

**Review by dnys1 — APPROVED** [review]

File: packages/amplify_api_plugin_interface/lib/src/types/pagination/paginated_result.dart (hunk 2 above)
Comment (ragingsquirrel3) [reply packages/amplify_api_plugin_interface/lib/src/types/pagination/paginated_result.dart]: done, thanks

**Pull request merged**